Define strong uuid-class types for table_id, table_schema_version and query_id

bhalevy opened the pull request:
We would like to define more distinct types
that are currently defined as aliases to utils::UUID to identify
resources in the system, like table id and schema
version id.

As with counter_id, the motivation is to restrict
the usage of the distinct types so they can be used
(assigned, compared, etc.) only with objects of
the same type.  Using with a generic UUID will
then require explicit conversion, that we want to
expose.

This series starts with cleaning up the idl header definition
by adding support for `import` and `include` statements in the idl-compiler.
These allow the idl header to become self-sufficient
and then remove manually-added includes from source files.
The latter usually need only the top level idl header
and it, in turn, should include other headers if it depends on them.

Then, a UUID_class template was defined as a shared boiler plate
for the various uuid-class.  First, we convert counter_id to use it,
rather than mimicking utils::UUID on its own.

On top of utils::UUID_class<T>, we define table_id,
table_schema_version, and query_id.

Following up on this series, we should define more commonly used
types like: host_id, streaming_plan_id, paxos_ballot_id.

Fixes #11207

bhalevy commented:
Cc @Deexie, task_manager::task_id should be based on utils::UUID_class when it becomes available.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1687/

psarna commented:
@bhalevy CI suggests that while making headers self-sufficient, you uncovered a preexisting sstring.hh dependency missing in per partition limiting, which was concealed by the fact that it included serializer.hh, which also provided string.hh

denesb commented on idl-compiler.py:
```diff
@@ -74,6 +74,30 @@ def ns_qualified_name(self):
+    def __repr__(self):
+        return self.__str__()
+
+class Include(ASTBase):
```
What is the difference between `import` and `include`?

denesb replied:
I see it down below, but we need to document it somewhere. Speaking of which, this whole IDL is woefully under-documented (or documentation is just very hard to discover). Not this series's fault of course.

bhalevy replied:
`include` just adds a respective `#include` directive in the generated `.dist.hh` header using the same parameter.

`import name` refers to another `name.idl.hh` "module" and adds `#include "name.dist.hh"` and  `#include "name.dist.impl.hh"` to the respective generated header files.

denesb commented on idl/idl_test.idl.hh:
```diff
@@ -82,3 +84,7 @@ struct const_template_arg_wrapper {
     std::vector<const_template_arg_wrapper<const simple_compound>> first;
 };
+
+struct frozen_object stub [[writable]] {
```
Maye move this to a separate patch.

bhalevy replied:
sure, np

denesb commented on db/batchlog_manager.cc:
```diff
@@ -31,13 +31,9 @@
-#include "serializer_impl.hh"
-#include "serialization_visitors.hh"
-#include "db/schema_tables.hh"
-#include "idl/uuid.dist.impl.hh"
```
Nice. I always hated the obscure error generated by these missing headers and how sensitive it all was to ordering.

bhalevy commented:
> @bhalevy CI suggests that while making headers self-sufficient, you uncovered a preexisting sstring.hh dependency missing in per partition limiting, which was concealed by the fact that it included serializer.hh, which also provided string.hh

I wonder why I haven't hit it when building locally.
I'll reproduce and fix.

denesb commented on db/hints/sync_point.hh:
```diff
@@ -39,5 +39,25 @@ struct sync_point {
+
+    // Sync point information for materialized view hints
+    db::hints::per_manager_sync_point_v1 mv_sp;
+};
```
Where were these before? And why weren't they needed?

bhalevy replied:
oddly enough, https://github.com/scylladb/scylladb/blob/527b34507910ba5606864468e81718f228257450/db/hints/sync_point.cc#L21

included `idl/hinted_handoff.idl.hh` itself.
This worked somehow, but the idl headers aren't really design to be included verbatim by the .cc sources.
The have a hidden assumption that the "real" data structures they represent are defined somewhere else.

With the added support for `include` and `import` statements in idl, it can no longer be included as it was.

I'll separate it to a preliminary patch to clarify what's going on here.

denesb commented on utils/UUID.hh:
```diff
@@ -206,7 +206,47 @@ inline std::strong_ordering uuid_tri_compare_timeuuid(bytes_view o1, bytes_view
+template <typename T>
+class UUID_class {
+    int64_t most_sig_bits;
+    int64_t least_sig_bits;
```
Why not store UUID instance?

bhalevy replied:
Because it is not trivially-default-constructible (it initializes the members to 0)
and so it can't use the serialization templates for trivial types.
To reduce the churn, I just followed the existing `counter_id` approach which is to mimic a UUID and provide conversion methods.

There's a static_assert that checks that these are POD types.

denesb commented on utils/UUID.hh:
```diff
@@ -206,7 +206,47 @@ inline std::strong_ordering uuid_tri_compare_timeuuid(bytes_view o1, bytes_view
+    int64_t most_sig_bits;
+    int64_t least_sig_bits;
+public:
+    UUID_class() = default;
```
What is the intended semantics of a default constructed `UUID_class`? Currently it just has two uninitialized fields and will compare equal/less/more to other instances depending on what garbage that memory happened to contain. I think we need a well defined state for a default constructed `UUID_class`.

bhalevy replied:
Right.  The intention is to keep it as uninitialized so it will be trivially default constructible (See https://en.cppreference.com/w/cpp/types/is_trivial)

denesb replied:
I think you can still initialize members as in `int64_t most_sig_bits = 0`. The constructor is trivial as long as it is not user-defined (i.e. until the compiler can definitely prove that it is not running any user logic).

denesb replied:
If that doesn't work we should chose safety over convenience. We can write a templated serializer that works for all UUID_class types easily.

bhalevy replied:
> I think you can still initialize members as in int64_t most_sig_bits = 0. The constructor is trivial as long as it is not user-defined (i.e. until the compiler can definitely prove that it is not running any user logic).

That was one of the first things I tried.  The compiler thinks differently than our intuition :)

> If that doesn't work we should chose safety over convenience. We can write a templated serializer that works for all UUID_class types easily.

I'll take a stab at it.

If this works, We should be able to define utils::UUID itself as derived from UUID_class as well.
So all uuid-classes would be symmetrical.

denesb replied:
I think we should still keep using composition instead of inheritance, to avoid accidental up-casting.

bhalevy replied:
The idea is not to derive UUID_class<T> from UUID, but that rather define UUID itself as a strong class
like `struct UUID : public UUID_class<UUID>`, and that should be immune to accidental type casting to other types derived from `UUID_class` due to the template type tag.

denesb replied:
I don't really follow. How will inheriting `UUID` from some type make it a strong type?

bhalevy replied:
never mind.  I tried the common ancestor approach and it is too complicated.
Next version has a UUID_class<T> base template that carries a utils::UUID member.

The culprit requiring trivial types was in bit_cast, but the requirement was too harsh
as proven by https://en.cppreference.com/w/cpp/numeric/bit_cast requires only `std::is_trivially_copyable_v`.

I'm fixing that along the way.

denesb commented on alternator/streams.cc:
```diff
@@ -74,8 +74,8 @@ struct rapidjson::internal::TypeHelper<ValueType, utils::UUID>
     : public from_string_helper<ValueType, utils::UUID>
 {};
 
-static db_clock::time_point as_timepoint(const utils::UUID& uuid) {
```
s/concrete type/strong type/

bhalevy replied:
ok

denesb commented on db/hints/manager.cc:
```diff
@@ -506,7 +506,7 @@ future<> manager::end_point_hints_manager::flush_current_hints() noexcept {
 
 class no_column_mapping : public std::out_of_range {
```
s/concrete type/strong type/

bhalevy replied:
ok

denesb commented on schema_fwd.hh:
```diff
@@ -31,3 +31,27 @@ public:
+//
+// When table_schema_version changes, schema_tables::calculate_schema_digest() should
+// also change when schema mutations are applied.
+class table_schema_version : public utils::UUID_class<table_schema_version> {
```
nice trick to avoid a dummy tag type

bhalevy replied:
inspired by `peering_sharded_service` :)

denesb commented on alternator/executor.cc:
```diff
@@ -3618,7 +3618,7 @@ static future<executor::request_return_type> do_query(service::storage_proxy& pr
         if (schema->clustering_key_size() > 0) {
```
s/concrete type/strong type/

bhalevy replied:
ok

denesb reviewed (changes requested):
Looks good, I think this is a very welcome change.
Had some comments abut the implementation.

tgrabiec commented:
We already have something like this as `tagged_id` in raft/internal.hh.

bhalevy commented:
> We already have something like this as `tagged_id` in raft/internal.hh.

Cool, I see now.  They are practically the same.
I'll convert it to to share the generic UUID_class template.

tgrabiec commented:
What's the advantage of this approach compared to the existing `tagged_id` approach? The latter is concise and non-intrusive to the UUID class.

bhalevy commented:
> What's the advantage of this approach compared to the existing `tagged_id` approach? The latter is concise and non-intrusive to the UUID class.

The `tagged_id` is good.  I'm preparing v2 based on it so that both raft::internal::tagged_id and the new types will be based on a common `utils::tagged_uuid<Tag>` template.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1697/

bhalevy commented:
In v2(dc46c19b1c)
- rebased + retested
- based on raft::internal::tagged_id as utils::tagged_uuid
- idl_test: add a struct depending on UUID
  - extracted as separate patch
- idl-compiler: add import and include statements
  - added documentation in `docs/dev/IDL.md`
- idl-compiler: include serialization impl and visitors in generated dist.impl.hh files
  - extracted as separate patch
- db/per_partition_rate_limit: tidy up headers self-sufficiency
  - resolves self-sufficiency issue hit in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1687/
- db: hints: sync_point: do not include idl definition file
  - extracted as separate patch
- Added patches for cleaning up bit_cast header file and overly restrictive is_trivial_v requirement:
  - bit_cast: use std::bit_cast
  - repair: delete unused include of utils/bit_cast.hh
  - utils: bit_cast: require TriviallyCopyable To
  - counters: counter_id delete requirement for triviality
- raft: migrate tagged_id definition to utils::tagged_uuid: provides infrastructure for the new strong types
  - utils: tagged_uuid: bool: reuse uuid::bool operator
  - utils: tagged_uuid: mark functions noexcept
- Following patches are now bases on utils::tagged_uuid and cleaned up:
  - counters: base counter_id of utils::tagged_uuid
  - counters: counter_id: use base class create_random_id
  - utils: tagged_uuid: rename to_uuid() to uuid()
  - schema, everywhere: define and use table_id as a strong type
  - schema, everywhere: define and use table_schema_version as a strong type
  - query-request, everywhere: define and use query_id as a strong type

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1698/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1699/

denesb commented on types/query_id.hh:
```diff
@@ -0,0 +1,13 @@
+
+#include "utils/UUID.hh"
+
+using query_id = utils::tagged_uuid<struct query_id_tag>;
```
`types/` directory is for storage system type system, not for random utility types we use. I think `query-request.hh` is a better place for this definition.

bhalevy replied:
ok, query-result.hh include query-request.hh so that could work.

bhalevy replied:
Done in v3

bhalevy commented on utils/bit_cast.hh:
```diff
@@ -9,18 +9,19 @@
+template <TriviallyCopyable To>
 To read_unaligned(const void* src) {
-    To dst;
+    uint8_t dst[sizeof(To)];
```
In hindsight we can just leave it as `To dst`.
Sine it's trivially copyable he function can just return the value as it did before.

bhalevy replied:
Done in v3

bhalevy commented on schema_fwd.hh:
```diff
@@ -19,3 +21,5 @@ class schema;
 
 using schema_ptr = seastar::lw_shared_ptr<const schema>;
+
+using table_id = utils::tagged_uuid<struct table_id_tag>;
```
It's bad to redefine the type in two places that may go out of sync.
Maybe we need to change schema_fwd.hh to schema_basic_types.hh.
FWIW, the definition of schema_ptr here is also not a forward declaration.

bhalevy replied:
Done in v3

bhalevy commented on schema_fwd.hh:
```diff
@@ -23,3 +24,21 @@ class schema_extension;
+
+inline table_schema_version reversed(table_schema_version v) noexcept {
+    return table_schema_version(utils::UUID_gen::negate(v.uuid()));
+}
```
The problem is that schem.hh is too heavy for most use cases that don't need the full schema definition.
Again, changing schem_fwd.hh to schema_basic_types.hh can satisfy both needs.

bhalevy replied:
Done in v3

bhalevy commented:
In v3 (6d80b8270):
- rebased + retested
- utils: bit_cast: require TriviallyCopyable To
  - keep `To dst` (do not use `uint8_t dst[]`)
- schema: rename schema_fwd.hh to schema_basic_types.hh
- query-request, everywhere: define and use query_id as a strong type
  - define in query-request.hh (do not introduce types/query_id.hh)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1701/

avikivity commented on docs/dev/IDL.md:
```diff
@@ -30,6 +32,17 @@ declaration, similar to C++.
+To import definitions from other IDL definition files, use the `import <other_mod_name>` statement.
+The idl compiler generates a respective `#include "<other_mod_name>.dist.hh"` directive in the generated `<mod_name>.dist.hh` file
+and a `#include "<other_mod_name>.dist.impl.hh"` directive in the generated `<mod_name>.dist.impl.hh` file
+
```
IDL was designed to mimic C++ syntax, so we can switch to C++ reflection when it becomes available (optimistically in C++26). So I'd rather not add new syntax, which conflicts with C++ module import.

Can we make #include detect that it's including idl? For example, #include "xyz.idl.hh" is importing an idl module.

bhalevy replied:
I don't know how

avikivity replied:
str.endswith(".idl.hh")?

bhalevy replied:
ah, so it will look like C pre-processor, but will do more.
that's possible.

avikivity commented on utils/UUID.hh:
```diff
@@ -228,4 +226,8 @@ struct hash<utils::UUID> {
+
+inline std::ostream& operator<<(std::ostream& out, const utils::UUID& uuid) {
+    return out << uuid.to_sstring();
 }
```
Why make it inline?

bhalevy replied:
it has to be inline when moved to the header file.
but actually this was done in an earlier version that tried to base utils::UUID itself and the base template, but I think this is no longer needed.

avikivity replied:
Why can't you have a declaration but not a definition?

bhalevy replied:
As I tried to explain above, I can.
It was a preparation patch from an older version that is no longer needed.

avikivity commented on counters.hh:
```diff
@@ -18,45 +18,17 @@
-class counter_id {
-    int64_t _least_significant;
-    int64_t _most_significant;
+class counter_id : public utils::tagged_uuid<counter_id> {
```
Why not: using counter_id = utils::tagged_uuid<class something>? Avoids a wrapper class.

bhalevy replied:
For `generate_random` below, but this is only a technical step for the patchset.
It is replaced in the next patch with `using ...`

avikivity commented on counters.hh:
```diff
@@ -18,16 +18,7 @@
-        return counter_id(utils::make_random_uuid());
-    }
-};
+using counter_id = utils::tagged_uuid<struct counter_id_tag>;
```
Oh.

avikivity commented on alternator/serialization.hh:
```diff
@@ -12,7 +12,7 @@
 #include <optional>
 #include "types.hh"
-#include "schema_fwd.hh"
+#include "schema_basic_types.hh"
```
I think this patch is just noise. Nobody complained that schema_fwd.hh wasn't 1000% accurate, everyone understood what it meant. Now they have to learn something new for no gain.

bhalevy replied:
Botond thought that adding table_id and table_schema_version to it are out of scope.
Therefore the rename serves to increase the scope of the header file.
I don't mind keeping it as is and add the definitions there.

We can define them just in schema.hh but the occasional users will have to include schema.hh that has a lot of bagage.  Maybe that's ok.

However, I do mind defining it the same way in two places, e.g both in schema.hh and schema_fwd.hh.
That's a very bad habbit.

avikivity replied:
Doesn't schema.hh include schema_fwd.hh?

bhalevy replied:
It definitely should have, but it didn't up until this series.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -85,9 +85,9 @@ api::timestamp_type system_keyspace::schema_creation_timestamp() {
+table_schema_version system_keyspace::generate_schema_version(::table_id table_id, uint16_t offset) {
     md5_hasher h;
-    feed_hash(h, table_id);
+    feed_hash(h, table_id.uuid());
```
We should have feed_hash(h, that_template) so this isn't needed.

bhalevy replied:
Yes.  This does the trick:
```
diff --git a/utils/UUID.hh b/utils/UUID.hh
index 1d86249135..8635c07a5f 100644
--- a/utils/UUID.hh
+++ b/utils/UUID.hh
@@ -254,6 +254,14 @@ struct appending_hash<utils::UUID> {
     }
 };
 
+template<typename Tag>
+struct appending_hash<utils::tagged_uuid<Tag>> {
+    template<typename Hasher>
+    void operator()(Hasher& h, const utils::tagged_uuid<Tag>& id) const noexcept {
+        appending_hash<utils::UUID>{}(h, id.uuid());
+    }
+};
+
 namespace std {
 template<>
 struct hash<utils::UUID> {
```

avikivity commented on idl/uuid.idl.hh:
```diff
@@ -19,3 +19,7 @@ class UUID final {
+
+class table_schema_version final {
+    utils::UUID uuid();
+};
```
table_id and table_schema_version should be defined near their users, not in a generic uuid module. However, it's probably too hairy to untangle so we can live with it for now.

bhalevy replied:
yes, it is used in multiple places and even with the support for 'import' it is hard to decide who's the lowest level user for it so it can be defined there.

avikivity replied:
The standard approach is to define their own modules. But let's not do it now.

bhalevy replied:
yes, I thought of `schema.idl.hh`, but we I agree it's an overkill for just those 2 basic types.

avikivity commented:
Looks good, left some minor comments. 

I think the tagged_uuid explicit constructor is not explicit enough, it's easy to call it by mistake. I suggest to change it to from_uuid(), but not in this series, it's long enough.

bhalevy commented:
In v4 (eed8124f4c):
- idl-compiler: add include statements
  - use `#include` for both cases
  - updated doc respectively
- dropped: "utils: uuid: define operator<< inline in header file"
- added "utils: uuid: define appending_hash<utils::tagged_uuid<Tag>>"
- added "schema: include schema_fwd.hh in schema.hh"
- dropped "schema: rename schema_fwd.hh to schema_basic_types.hh"
  - use schem_fwd.hh to define both `table_id` and `table_schema_version` using `utils::tagged_uuid`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1708/

denesb reviewed (approved):
I'm not happy about `schema_fwd.hh` becoming a second `schema.hh`, housing a random subset of the schema-related types. In my world, a `*_fwd.hh` header is not a file I ever have to open to look for a definition: it just contains short re-declarations of stuff already declared fully elsewhere. Yes, in some cases there is no shorter form of declarations, so some stuff may be repeated.
But it looks like others don't mind this, furthermore this practice already has precedents, so I won't object.

denesb commented:
Conflicts, please rebase.

bhalevy commented:
c71ef330b2 rebased on top of ba42852b0e1d2bec677e60a863c505a637884e04

annastuchlik commented:
@scylladb/scylla-maint Please backport to 5.1

bhalevy commented:
> @scylladb/scylla-maint Please backport to 5.1

why?

denesb commented:
We are preparing for per-version documentation.

bhalevy commented:
> We are preparing for per-version documentation.

But this isn't a documentation change

denesb commented:
This might be a mistake then, @annastuchlik ?
